### PR TITLE
Add the permutation related support as the extension for asp lib.

### PR DESCRIPTION
**Pull request by ChongyuNVIDIA**
Permutation search is the extension to the ASP (Automatic SParsity) lib. It can help to reduce the accuracy drop due to structured sparse pruning.

Have done E2E checking with various TorchVision networks.
- PyTorch >= 1.8.0
- TorchVision >= 0.9.0

Have offline reviewed with @thorjohnsen 

Submit the Pull Request to @thorjohnsen @ptrblck

### CODE REVIEW

File: apex/contrib/sparsity/permutation_search_kernels/permutation_utilities.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            # Method No.2, More robust
+            import sysconfig
+            lib_dir = sysconfig.get_config_var("LIBDEST")    # e.g. '/opt/conda/lib/python3.8'
+            #lib_dir = sysconfig.get_config_var("BINLIBDEST")    # e.g. '/opt/conda/lib/python3.8'
+            #print(lib_dir)
+            ext_suffix = sysconfig.get_config_var('EXT_SUFFIX')    # e.g. '.cpython-38-x86_64-linux-gnu.so'
+            #print(ext_suffix)
+            cuda_kernel_name = 'permutation_search_cuda' + ext_suffix
+            #print(cuda_kernel_name)    # e.g. permutation_search_cuda.cpython-38-x86_64-linux-gnu.so
+            cuda_kernel_path = os.path.join(lib_dir, 'site-packages', cuda_kernel_name)    # e.g. /opt/conda/lib/python3.8/site-packages/permutation_search_cuda.cpython-38-x86_64-linux-gnu.so
```
Comment (ptrblck) [inline apex/contrib/sparsity/permutation_search_kernels/permutation_utilities.py:50]: This is an invalid path, as e.g. the installation in a conda base environment would point to:
```
/opt/conda/lib/python3.8/site-packages/apex-0.1-py3.8-linux-x86_64.egg/permutation_search_cuda.cpython-38-x86_64-linux-gnu.so
```

**Review by crcrpar — COMMENTED** [review]
While toy_problem.py seems to work well but I'm not sure whether or not it's allowed to ignore the failing test scripts.

```
root@nvdl-smc-07:/mkozuki/repositories/apex_1# python apex/contrib/sparsity/test/checkpointing_test_part1.py  >> /dev/null
Traceback (most recent call last):
  File "apex/contrib/sparsity/test/checkpointing_test_part1.py", line 94, in <module>
    main(args)
  File "apex/contrib/sparsity/test/checkpointing_test_part1.py", line 58, in main
    ASP.compute_sparse_masks()
  File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/asp.py", line 247, in compute_sparse_masks
    mask.set_(cls.__calculate_mask(p))
  File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/asp.py", line 89, in create_mask_from_pattern
    return create_mask(param, mask_calculator).bool()
  File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/sparse_masklib.py", line 161, in create_mask
    mask = func(t, density)
  File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/sparse_masklib.py", line 141, in m4n2_2d_best
    return mn_2d_best(mat, 4, 2)
  File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/sparse_masklib.py", line 128, in mn_2d_best
    mat = reshape_2d(matrix,m,m).abs()
NameError: name 'reshape_2d' is not defined
```

**Comment by ptrblck** [comment]
Discussed with @ChongyuNVIDIA: failures are unrelated to this PR, merging.

**Comment by Triple-L** [comment]
> While toy_problem.py seems to work well but I'm not sure whether or not it's allowed to ignore the failing test scripts.
> 
> ```
> # python apex/contrib/sparsity/test/checkpointing_test_part1.py  >> /dev/null
> Traceback (most recent call last):
>   File "apex/contrib/sparsity/test/checkpointing_test_part1.py", line 94, in <module>
>     main(args)
>   File "apex/contrib/sparsity/test/checkpointing_test_part1.py", line 58, in main
>     ASP.compute_sparse_masks()
>   File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/asp.py", line 247, in compute_sparse_masks
>     mask.set_(cls.__calculate_mask(p))
>   File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/asp.py", line 89, in create_mask_from_pattern
>     return create_mask(param, mask_calculator).bool()
>   File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/sparse_masklib.py", line 161, in create_mask
>     mask = func(t, density)
>   File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/sparse_masklib.py", line 141, in m4n2_2d_best
>     return mn_2d_best(mat, 4, 2)
>   File "/mkozuki/repositories/apex_1/apex/contrib/sparsity/sparse_masklib.py", line 128, in mn_2d_best
>     mat = reshape_2d(matrix,m,m).abs()
> NameError: name 'reshape_2d' is not defined
> ```

Hi crcrpar
do you know how to fix this bug now?

**Pull request merged**